lib/Sema/CSSimplify.cpp

```diff
@@ -5785,6 +5785,43 @@ ConstraintSystem::matchTypes(Type type1, Type type2, ConstraintKind kind,
                                    ConversionRestrictionKind::PointerToPointer);
               }
             }
+
```

Comment:
It looks to me like the ~150 lines inside the `getAnyPointerElementType` check would be nicely separable in a function (as a separate cleanup)

Reply:
Or, at least the part that's specific to `ConstraintKind::ArgumentConversion`

Reply:
Sure, I think it would be a good exercise to try to factor all the logic into restriction kinds e.g. `simplifyRestriction(...)` just like we do with constraints. Something I wanted to do but never had time...